File: designPatterns/Behavioral_patterns/Iterator/javascript/WordsCollection.js
const AlphabeticalOrderIterator = require("./AlphabeticalOrderIterator");

/**
 * Concrete Collections provide one or several methods for retrieving fresh
 * iterator instances, compatible with the collection class.
 */
class WordsCollection {
  constructor() {
    this.items = [];
  }

  getItems() {
    return this.items;
  }

  getCount() {
    return this.items.length;
  }

  addItem(item) {
    this.items.push(item);
  }

  getIterator() {
    return new AlphabeticalOrderIterator(this);
  }

  getReverseIterator() {
    return new AlphabeticalOrderIterator(this, true);
  }
}

module.exports = WordsCollection;

File: designPatterns/Behavioral_patterns/Iterator/python/WordsCollection.py
from AlphabeticalOrderIterator import AlphabeticalOrderIterator


class WordsCollection:
    def __init__(self):
        self.items = []

    def get_items(self):
        return self.items

    def get_count(self):
        return len(self.items)

    def add_item(self, item):
        self.items.append(item)

    def get_iterator(self):
        return AlphabeticalOrderIterator(self)

    def get_reverse_iterator(self):
        return AlphabeticalOrderIterator(self, True)
